### PR TITLE
Added comment resolver (fixes #497)

### LINKED ISSUE #497: Add comments resolver for wowrk item service

### PR DESCRIPTION
**Pull request by sanbornsen**
Features - 

- Supporting new comment model and related link in comment
- Quick loading work item details page with existing data
- Resolve comments every time details page gets loaded
- Moved comment services to work item service
- Maintaining a big list of work item and relationships for easy access

### CODE REVIEW

**Comment by sanbornsen** [comment]
If you are reviewing, please go through every single functionality manually and thoroughly.

**Comment by aslakknutsen** [comment]
I'm getting a 404 on http://localhost:8088/settings.json ?

Not sure if that is new with this one or not but..

**Comment by aslakknutsen** [comment]
And I see the order of the comments has changed? They used to be last first as proposed in the original designs around comments. Intentional?

**Comment by aslakknutsen** [comment]
But Boy does it feel snappier! Great work.

Would generally love some kind of 'loading indicator', both for when details refresh and when 'next page' is fetched.

**Comment by sanbornsen** [comment]
@aslakknutsen 
- That 404 is coming from some obsidian stuff, it's been there for a while now. It's there in production also.
- Yeah, comment is my fault, I'll fix that. 
- I am waiting for next UX meetup to discuss about that. @michaelkleinhenz and me had an interesting discussion yesterday about that. We'll load the data and make a call in background for the details page. The old available data will be read-only and showing some kind of loader until the new data comes in. For now if this change goes in, the details page will open in no time with the locally available data.

Thanks :)

File: src/app/work-item/work-item-list/work-item-detail/work-item-comment/work-item-comment.component.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
+import { WorkItemService } from './../../../work-item.service';
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-comment/work-item-comment.component.ts]: system.js path style

File: src/app/work-item/work-item.service.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import { Injectable } from '@angular/core';
+import { Comments, Comment, CommentPost } from './../models/comment';
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: system.js path style

File: src/app/work-item/work-item.service.ts [hunk 3]
```diff
@@ -100,12 +125,83 @@ export class WorkItemService {
...
+   * New item will be added to the list
+   */
+  updateWorkItem(wItems: WorkItem[]): void {
+    wItems.forEach((wItem) => {
+      if (wItem.id in this.workItemIdIndexMap) {
+        this.workItems[this.workItemIdIndexMap[wItem.id]].attributes = 
+          cloneDeep(wItem.attributes);
+      } else {
+        this.workItems
+          .splice(this.workItems.length, this.workItems.length, wItem)
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: missing semicolon

File: src/app/work-item/work-item.service.ts [hunk 4]
```diff
@@ -222,10 +391,21 @@ export class WorkItemService {
...
         // removing deleted item from the local list
         this.workItems.splice(deletedItemIndex, 1);
+        // Re build the workItem ID-Index map
+        this.buildWorkItemIdIndexMap();
       })
       .catch(this.handleError);
   }
 
+   /**
+   * Usage: This method create a new item
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: this just hurts my sensibilities, sorry but could please align the asterisk in the comment block

File: src/app/work-item/work-item.service.ts [hunk 5]
```diff
@@ -54,26 +66,35 @@ export class WorkItemService {
       .get(url, { headers: this.headers })
       .toPromise()
       .then(response => {
+        // Build the user - id map
         this.buildUserIdMap();
+        let wItems: WorkItem[];
         if (process.env.ENV == 'inmemory') {
           // Exclusively for in memory testing
-          this.workItems = response.json().data as WorkItem[];
-          this.workItems = this.workItems.reverse();
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: Why are you removing the reverse? How is that related to the comments?

File: src/app/work-item/work-item.service.ts [hunk 6]
```diff
@@ -54,26 +66,35 @@ export class WorkItemService {
       .get(url, { headers: this.headers })
       .toPromise()
       .then(response => {
+        // Build the user - id map
         this.buildUserIdMap();
+        let wItems: WorkItem[];
         if (process.env.ENV == 'inmemory') {
           // Exclusively for in memory testing
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: Is this really exclusively for in memory?

File: src/app/work-item/work-item.service.ts [hunk 7]
```diff
@@ -100,12 +125,83 @@ export class WorkItemService {
     }
   }
 
+  /**
+   * Usage: This method gives a single work item by ID. 
+   * If the item is locall available then it just resolves the comments
+   * else it fethces the item from cloud and then resolve the comments
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: typos: locall, fethces, from cloud

File: src/app/work-item/work-item.service.ts [hunk 8]
```diff
@@ -54,26 +66,35 @@ export class WorkItemService {
...
         return this.workItems;
       })
       .catch (this.handleError); 
   }
 
+  /**
+   * This function is called from next page onwards in the scroll
+   * It does pretty much same as the getWorkItems function
+   */
+
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: It is just a style thing but usually the JSDoc that goes with a function or class is directly above the function or class with no space or blank line. This visually ties them together.

File: src/app/work-item/work-item.service.ts [hunk 9]
```diff
@@ -163,10 +261,57 @@ export class WorkItemService {
...
+   * this is eventually be deprecated once work item linking is re-worked
+   */
+
   getLocallySavedWorkItems(): Promise<any> {
     return Promise.resolve(this.workItems);
   }
 
+  /**
+   * Usage: This method is to resolve the comments for a work item 
+   * This ethod is only called when a single item is fetched for the
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: typo: ethod

for reference, in JavaScript methods are called functions

File: src/app/work-item/work-item.service.ts [hunk 10]
```diff
@@ -163,10 +261,57 @@ export class WorkItemService {
...
 
+  /**
+   * Usage: This method is to resolve the comments for a work item 
+   * This ethod is only called when a single item is fetched for the 
+   * details page.
+   * 
+   * @Input: WorkItem - wItem
+   */
+
+  resolveComments(wItem: WorkItem): void {
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: It appears this function is only called for in memory operations. Is this correct?

File: src/app/work-item/work-item.service.ts [hunk 11]
```diff
@@ -143,13 +242,12 @@ export class WorkItemService {
...
 
-  getUserById(userId: string): User {
-    if (userId in this.userIdMap) {
-      return this.userIdMap[userId];
-    } else {
-      return null;
-    }
-  }
+  /**
+   * Usage: Build a ID-User map to fast access of list of users
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: typo: "to fast access" I think it is missing a verb

**Comment by joshuawilson** [comment]
There is a failure in the work-item-detail.component.spec.ts starting at line 84.

File: src/app/work-item/work-item.service.ts (hunk 2 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Ha ha, about this, do you know VS code does that automatically when I try to import a package. I didn't even write this line, it's added automatically by VS Code and they follow this path style.

File: src/app/work-item/work-item.service.ts (hunk 5 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Just to reduce a line, that's all.

Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Reversing has been there from the begging. Once we have ordered data from backend we can make it cleaner.

File: src/app/work-item/work-item.service.ts (hunk 2 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: I'll change it anyway.

Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: ok, then I'll stop ask you to change it. I'm tired of fighting VS Code.

File: src/app/work-item/work-item.service.ts (hunk 6 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Yeah, true. We don't need that anymore. Just bear with it for couple more days. We are going to get rid off all in-memory stuffs. So, I just left it as it is fir now.

File: src/app/work-item/work-item.service.ts [hunk 12]
```diff
@@ -163,10 +261,57 @@ export class WorkItemService {
...
+    if (userId in this.userIdMap) {
+      return this.userIdMap[userId];
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * This is to fetch locally fetched work items
+   * this is eventually be deprecated once work item linking is re-worked
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Damn !! my English !!

File: src/app/work-item/work-item.service.ts (hunk 10 above)
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: No, check the else part.
